libdisco/asymmetric.go

```diff
@@ -59,85 +62,173 @@ func dh(keyPair KeyPair, publicKey [32]byte) (shared [32]byte) {
 	return
 }
 
-// uses deterministic Schnorr with strobe
-type SigningKeyPair struct {
+// SigningKeypair uses deterministic Schnorr with strobe
+type SigningKeypair struct {
 	SecretKey ristretto.Scalar
-	PublicKey ristretto.Point
+	PublicKey ristretto.Element
 }
 
-//
-func GenerateSigningKeyPair() SigningKeyPair {
-	// Generate an El'Gamal keypair
-	var secretKey ristretto.Scalar
-	var publicKey ristretto.Point
+// Signature represents a schnorrkel signature
+type Signature struct {
+	R ristretto.Element
+	S ristretto.Scalar
+}
+
+// Decode a schnorrkel signature from a bytearray.
+// ref: https://github.com/w3f/schnorrkel/blob/master/src/sign.rs#L100
+func (s *Signature) Decode(sigBytes [64]byte) error {
+	err := s.R.Decode(sigBytes[:32])
+	if err != nil {
+		return err
+	}
+	sigBytes[63] &= 127
+	err = s.S.Decode(sigBytes[32:])
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Encode a signature as a bytearray.
+// see: https://github.com/w3f/schnorrkel/blob/master/src/sign.rs#L77
+func (s *Signature) Encode() [64]byte {
+	var sigBytes [64]byte
+
+	rBytes := s.R.Encode(nil)
+	copy(sigBytes[:32], rBytes)
+
+	sBytes := s.S.Encode(nil)
+	copy(sigBytes[32:], sBytes)
+
+	sigBytes[63] |= 128
+
+	return sigBytes
+}
 
-	secretKey.Rand()                     // generate a new secret key
-	publicKey.ScalarMultBase(&secretKey) // compute public key
+// GenerateSigningKeypair for schnorr signatures.
+func GenerateSigningKeypair() (SigningKeypair, error) {
 
-	return SigningKeyPair{secretKey, publicKey}
+	var sigpair SigningKeypair
+	// Generate a schnorr keypair
+	var publicKey ristretto.Element
+
+	secretKey, err := newRandomScalar()
+	if err != nil {
+		return sigpair, err
+	}
+	sigpair.PublicKey = *publicKey.ScalarBaseMult(&secretKey)
+	sigpair.SecretKey = secretKey
+
+	return sigpair, nil
 }
 
-func (kp SigningKeyPair) ExportPublicKey() string {
-	return hex.EncodeToString(kp.SecretKey.Bytes()) + hex.EncodeToString(kp.PublicKey.Bytes())
+// String returns a hexstring encoding of the signing keypair.
+func (kp SigningKeypair) String() string {
+	return hex.EncodeToString(kp.SecretKey.Encode(nil)) + hex.EncodeToString(kp.PublicKey.Encode(nil))
 }
 
-//
-func (kp SigningKeyPair) Sign(message []byte) []byte {
-	// 1. deterministic ephemeral k; r=g^k
-	var kBytes [32]byte
-	copy(kBytes[:], Hash(append(kp.SecretKey.Bytes(), message...), 32))
+// Sign a message using a deterministic nonce
+func (kp SigningKeypair) Sign(message []byte) Signature {
```

Comment:
why not returning `[]byte` directly?

Reply:
There's a decoding and encoding that includes bitflipping so I thought it would be better to make signatures a single component and implement the serialization functions independantly.. make it more maintainable.